File: Swift2/Sources/Swift2/ValueSequencer.swift

public class CValueSequencer {

	init(nInnerSequencers: Int, nValues: Int, nValue: Int) {
		m_nInnerSequencers = nInnerSequencers
		m_nValues = nValues
		//m_pvsParent = null
		//m_pvsInner = null
		//m_nValue = nValue
		//SetValue(nValue: nValue)
	}

	public var Value: Int {
		get {return m_nValue}
		set(value) {SetValue(nValue: value)}
	}
	
	public var Sequence: Int {
		get {return m_nSequence}
		set(value) {SetValue(nValue: FindValue(nSequence: value))}
	}
	
	public var ValueStr: String {
		get { return String(Sequences.nHexagramSequences[0][Value])}
	}
	
	public var SequenceStr: String {
		get	{return String(Sequence + 1)}
	}
	
	public var Label: String {
		get { return GetLabel() }
	}
	
	public var IsMoving: Bool {
		get { return GetMoving() }
	}
	
	@discardableResult
	public func First() -> CValueSequencer {
		SetValue(nValue: GetFirstSequence())
		UpdateInnerValues()
		UpdateOuterValues()
		return self
	}

	@discardableResult
	public func Previous(bRatio: Bool = false) -> CValueSequencer {
		m_nRatio = m_nRatio - 1
		if !bRatio || (m_nRatio >= 0) {
			SetValue(nValue: GetPreviousSequence(nSequence: m_nSequence))
			UpdateInnerValues()
			UpdateOuterValues()
		}
		return self
	}

	@discardableResult
	public func Next(bRatio: Bool = false) -> CValueSequencer {
		m_nRatio = m_nRatio - 1
		if !bRatio || (m_nRatio >= 0) {
			SetValue(nValue: GetNextSequence(nSequence: m_nSequence))
			UpdateInnerValues()
			UpdateOuterValues()
		}
		return self
	}

	@discardableResult
	public func Last() -> CValueSequencer {
		SetValue(nValue: GetLastSequence())
		UpdateInnerValues()
		UpdateOuterValues()
		return self
	}

	@discardableResult
	public func Inverse() -> CValueSequencer { return self }

	@discardableResult
	public func Opposite() -> CValueSequencer {
		SetValue(nValue: ~m_nValue & (m_nValues - 1))
		UpdateInnerValues()
		UpdateOuterValues()
		return self
	}

	@discardableResult
	public func Transverse() -> CValueSequencer { return self }
	@discardableResult
	public func Nuclear() -> CValueSequencer { return self }
	@discardableResult
	public func Move() -> CValueSequencer { return self }
	@discardableResult
	public func Young() -> CValueSequencer { return self }
	@discardableResult
	public func Old() -> CValueSequencer { return self }

	@discardableResult
	public func Update() -> CValueSequencer {
		SetValue(nValue: m_nValue)
		UpdateInnerValues()
		UpdateOuterValues()
		return self
	}

	public func UpdateInnerValues() {}
	public func UpdateOuterValues() {}

	public func SetParent(pvsParent: CValueSequencer) { m_pvsParent = pvsParent }
	//public func GetParent() -> CValueSequencer { return m_pvsParent }
	
	//public func GetChild(nIndex: Int) -> CValueSequencer {
	//	return m_pvsInner[nIndex]
	//}
	
	@discardableResult
	private func SetValue(nValue: Int) -> CValueSequencer {
		if (nValue >= 0) && (nValue < m_nValues) {
			m_nValue = nValue
			m_nSequence = m_nSequences[GetCurrentSequence()][nValue]
			m_nRatio = m_nRatios[GetCurrentRatio()][nValue]
		}
		return self
	}

	private func FindValue(nSequence: Int) -> Int {
		for i in 0 ... m_nValues - 1 {
			if m_nSequences[GetCurrentSequence()][i] == nSequence {
				return i
			}
		}
		return -1
	}
	
	private func GetFirstSequence() -> Int {
		return FindValue(nSequence: 0)
	}

	private func GetPreviousSequence(nSequence: Int) -> Int {
		if nSequence > 0 {
			let s = nSequence - 1
			return FindValue(nSequence: s)
		}
		return GetLastSequence()
	}

	private func GetNextSequence(nSequence: Int) -> Int	{
		if nSequence < m_nValues - 1 {
			var s = nSequence + 1
			s = FindValue(nSequence: s)
			if s != -1 {
				return s
			}
		}
		return GetFirstSequence()
	}

	private func GetLastSequence() -> Int {
		var nSequence1 = 0
		var nSequence2 = m_nValues
		while nSequence2 > -1 {
			nSequence2 = nSequence2 - 1
			nSequence1 = FindValue(nSequence: nSequence2)
			if nSequence1 != -1 {
				return nSequence1
			}
		}
		return -1
	}
	
	public func GetLabel() -> String { return "" }
	public func GetMoving() -> Bool { return false }
	
	public func GetCurrentSequence() -> Int { return 0 }
	public func GetCurrentRatio() -> Int { return 0 }
	public func GetCurrentLabel() -> Int { return 0 }

	private var m_nValue = 0
	private var m_nSequence: Int = 0
	private var m_nRatio: Int = 0
	
	private var m_nInnerSequencers: Int = 0
	private var m_nValues: Int = 0
	
	var m_nSequences: [[Int]] = [[]]
	var m_nRatios: [[Int]] = [[]]

	var m_pvsParent: CValueSequencer!
	var  m_pvsInner: [CValueSequencer] = []
}

File: value_sequencer_py/value_sequencer.py
from __future__ import annotations

from typing import List, Optional

from sequences import Sequences


class CValueSequencer:
    """Base class for value sequencers."""

    def __init__(self, n_inner_sequencers: int, n_values: int, n_value: int) -> None:
        self._nInnerSequencers = n_inner_sequencers
        self._nValues = n_values
        self._nValue = n_value
        self._nSequence = 0
        self._nRatio = 0
        self._nSequences: List[List[int]] = []
        self._nRatios: List[List[int]] = []
        self._pvsParent: Optional[CValueSequencer] = None
        self._pvsInner: List[CValueSequencer] = []

    @property
    def value(self) -> int:
        return self._nValue

    @value.setter
    def value(self, value: int) -> None:
        self.set_value(value)

    @property
    def sequence(self) -> int:
        return self._nSequence

    @sequence.setter
    def sequence(self, sequence: int) -> None:
        self.set_value(self.find_value(sequence))

    @property
    def value_str(self) -> str:
        sequences = Sequences.nHexagramSequences
        if sequences and self.value < len(sequences[0]):
            return str(sequences[0][self.value])
        return str(self.value)

    @property
    def sequence_str(self) -> str:
        return str(self.sequence + 1)

    @property
    def label(self) -> str:
        return self.get_label()

    @property
    def is_moving(self) -> bool:
        return self.get_moving()

    def first(self) -> "CValueSequencer":
        self.set_value(self.get_first_sequence())
        self.update_inner_values()
        self.update_outer_values()
        return self

    def previous(self, ratio: bool = False) -> "CValueSequencer":
        if not ratio or self._decrement_ratio():
            self.set_value(self.get_previous_sequence(self._nSequence))
            self.update_inner_values()
            self.update_outer_values()
        return self

    def next(self, ratio: bool = False) -> "CValueSequencer":
        if not ratio or self._decrement_ratio():
            self.set_value(self.get_next_sequence(self._nSequence))
            self.update_inner_values()
            self.update_outer_values()
        return self

    def last(self) -> "CValueSequencer":
        self.set_value(self.get_last_sequence())
        self.update_inner_values()
        self.update_outer_values()
        return self

    def inverse(self) -> "CValueSequencer":
        return self

    def opposite(self) -> "CValueSequencer":
        self.set_value(~self._nValue & (self._nValues - 1))
        self.update_inner_values()
        self.update_outer_values()
        return self

    def transverse(self) -> "CValueSequencer":
        return self

    def nuclear(self) -> "CValueSequencer":
        return self

    def move(self) -> "CValueSequencer":
        return self

    def young(self) -> "CValueSequencer":
        return self

    def old(self) -> "CValueSequencer":
        return self

    def update(self) -> "CValueSequencer":
        self.set_value(self._nValue)
        self.update_inner_values()
        self.update_outer_values()
        return self

    def update_inner_values(self) -> None:
        pass

    def update_outer_values(self) -> None:
        pass

    def set_parent(self, parent: "CValueSequencer") -> None:
        self._pvsParent = parent

    def get_parent(self) -> Optional["CValueSequencer"]:
        return self._pvsParent

    def get_child(self, index: int) -> Optional["CValueSequencer"]:
        return self._pvsInner[index] if 0 <= index < len(self._pvsInner) else None

    def set_value(self, value: int) -> "CValueSequencer":
        if 0 <= value < self._nValues:
            self._nValue = value
            sequence_row = self._nSequences[self.get_current_sequence()]
            if value < len(sequence_row):
                self._nSequence = sequence_row[value]
            ratio_row = self._nRatios[self.get_current_ratio()]
            if value < len(ratio_row):
                self._nRatio = ratio_row[value]
        return self

    def find_value(self, sequence: int) -> int:
        sequence_row = self._nSequences[self.get_current_sequence()]
        for i in range(self._nValues):
            if sequence_row[i] == sequence:
                return i
        return -1

    def get_first_sequence(self) -> int:
        return self.find_value(0)

    def get_previous_sequence(self, sequence: int) -> int:
        if sequence > 0:
            return self.find_value(sequence - 1)
        return self.get_last_sequence()

    def get_next_sequence(self, sequence: int) -> int:
        if sequence < self._nValues - 1:
            value = self.find_value(sequence + 1)
            if value != -1:
                return value
        return self.get_first_sequence()

    def get_last_sequence(self) -> int:
        sequence = self._nValues
        while sequence > -1:
            found = self.find_value(sequence - 1)
            if found != -1:
                return found
            sequence -= 1
        return -1

    def get_label(self) -> str:
        return ""

    def get_moving(self) -> bool:
        return False

    def get_current_sequence(self) -> int:
        return 0

    def get_current_ratio(self) -> int:
        return 0

    def get_current_label(self) -> int:
        return 0

    def _decrement_ratio(self) -> bool:
        self._nRatio -= 1
        return self._nRatio == 0
